nisapi/clean.py

```diff
@@ -0,0 +1,253 @@
+                demographic_value=pl.when(
+                    pl.col("demographic_type") == pl.lit("overall")
+                )
+                .then(pl.lit("overall"))
+                .otherwise(pl.col("demographic_value"))
+            )
+        )
+
+        # there should now be no nulls in any column
+        assert clean.null_count().pipe(sum).item() == 0
+
+        # Remove duplicate rows
+        clean = clean.filter(clean.is_duplicated().not_())
+
+        # Find the rows that are *almost* duplicates: there are some rows that
+        # have nearly duplicate values
+        clean = clean.pipe(remove_near_duplicates)
+
+        # Verify that indicator type "up-to-date" has only one value ("yes")
+        assert clean.filter(pl.col("indicator_type") == pl.lit("up-to-date")).pipe(
+            col_values_in, "indicator_value", ["yes"]
+        )
+
+        # check that "Yes" and "Received a vaccination" are the same thing, so that
+        # we can drop "Up to Date"
+        assert (
+            clean.filter(
+                pl.col("indicator_value").is_in(["yes", "received a vaccination"])
+            )
+            .drop("indicator_type")
+            .pivot(on="indicator_value", values=["estimate", "ci_half_width_95pct"])
+            .select(
+                (
+                    (
+                        pl.col("estimate_yes")
+                        == pl.col("estimate_received a vaccination")
+                    )
+                    & (
+                        pl.col("ci_half_width_95pct_yes")
+                        == pl.col("ci_half_width_95pct_received a vaccination")
+                    )
+                ).all()
+            )
+            .item()
+        )
+
+        clean = clean.filter(
+            pl.col("indicator_type") == pl.lit("4-level vaccination and intent")
+        )
+
+        return clean
+
+    validate(clean)
+    return clean
+
+
+def rename_indicator_columns(df: pl.DataFrame) -> pl.DataFrame:
+    """
+    Make "indicator" follow the same logic as "geography" and
+    "demographic", with "type" and "value" columns
+    """
+    return df.rename(
+        {
+            "geographic_level": "geographic_type",
+            "geographic_name": "geographic_value",
+            "demographic_level": "demographic_type",
+            "demographic_name": "demographic_value",
+            "indicator_label": "indicator_type",
+            "indicator_category_label": "indicator_value",
+        }
+    )
+
+
+def col_values_in(df: pl.DataFrame, col: str, values: str) -> bool:
+    """All values of `df` in column `col` are in `values`?"""
+    return df[col].is_in(values).all()
+
+
+def remove_near_duplicates(df: pl.DataFrame, tolerance: float = 1e-3) -> pl.DataFrame:
+    value_cols = ["estimate", "ci_half_width_95pct"]
+    group_cols = set(df.columns) - set(value_cols)
+
+    nearly_dup_rows = df.drop(value_cols).is_duplicated()
+
+    # assert that there are only 2 of each of these
```

Comment:
Does this mean that you have already checked both data sets manually, and you know there are only 2 near-duplicates each time they occur? Because if there ever were 3+ near-copies of the same row, it seems we'd want to still get rid of all but one.